aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,248 @@
+
+        on_new_epoch(validator_set);
+        Vector::push_back(&mut validator_set.validators, addr);
+    }
+
+    /// Initiate by the validator info owner.
+    fun leave_validator_set(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+
+        let (exist, index) = Vector::index_of(&validator_set.validators, &addr);
+        assert!(exist, 0);
+        on_new_epoch(validator_set);
+        Vector::swap_remove(&mut validator_set.validators, index);
+        assert!(Vector::length(&validator_set.validators) > 0, 0);
+    }
+
+    /// Triggers when validator set changes or after certain time period (in block prologue).
+    fun on_new_epoch(validator_set: &mut ValidatorSet) acquires ValidatorInfo {
+        let current_time = Timestamp::now_seconds();
+        assert!(validator_set.last_update_time_secs + MINIMUM_RECONFIG_PERIOD > current_time, 0);
+        validator_set.last_update_time_secs = current_time;
+        let i = 0;
+        let len = Vector::length(&validator_set.validators);
+        while (i < len) {
+            let addr = *Vector::borrow(&validator_set.validators, i);
+            update_stake_pool(addr);
+            i = i + 1;
+        };
+        // Reconfiguration::reconfigure();
+        // do we remove validators without enough stakes?
+    }
+
+    /// Update individual validator's stake pool
+    /// 1. distribute rewards to active/pending_inactive delegations
+    /// 2. process pending_active, pending_inactive correspondingly
+    /// 3. update the current stake
+    fun update_stake_pool(addr: address) acquires ValidatorInfo {
+        let stake_pool = &mut borrow_global_mut<ValidatorInfo>(addr).stake_pool;
+        distribute_reward( &mut stake_pool.active);
+        distribute_reward( &mut stake_pool.pending_inactive);
+        // move pending_active to active
+        append(&mut stake_pool.active, &mut stake_pool.pending_active);
+        // move pending_inactive to inactive
+        append(&mut stake_pool.inactive, &mut stake_pool.pending_inactive);
+        let current_stake = 0;
+        let i = 0;
+        let len = Vector::length(&stake_pool.active);
+        while (i < len) {
+            current_stake = current_stake + TestCoin::value(&Vector::borrow(&stake_pool.active, i).coin);
+            i = i + 1;
+        };
+        stake_pool.current_stake = current_stake;
+    }
+
+    /// Mint the reward and add to the delegation based on some formula
+    fun distribute_reward(v: &mut vector<Delegation>) {
+        let i = 0;
+        let len = Vector::length(v);
+        while (i < len) {
+            let d = Vector::borrow_mut(v, i);
+            let reward = TestCoin::zero(); // mint some coins based on delegation, timestamp, maybe also totaly stakes
+            TestCoin::merge(&mut d.coin, reward);
+        };
```

Comment:
`i` doesn't seem like it's incremented here;

Reply:
classical move!